### PR TITLE
Support for complex weight initializers

### LINKED ISSUE #17097: Feature request: complex support in initializers
AFAIK no initializers work with tf.complex64 and tf.complex128 (is there a mathematical reason for this?).

My current workaround is to call them twice (once for the real part and once for the imaginary), as
```py
def complex_initializer(base_initializer):
    f = base_initializer()

    def initializer(*args, dtype=tf.complex64, **kwargs):
        real = f(*args, **kwargs)
        imag = f(*args, **kwargs)
        return tf.complex(real, imag)

    return initializer


tf.get_variable(
        name='my_complex_variable',
        shape=[1],
        dtype=tf.complex64,
        initializer=complex_initializer(tf.random_normal_initializer))
```
which seems to work fine so I'm wondering why this isn't built-in already?

### PR DESCRIPTION
**Pull request by DarshanDeshpande**
Adds support for complex weights to tf.keras's initializers. I am currently experiencing problems with Bazel on my local machine, so I was unable to run the tests, but only minor changes have been made to them, so they should, ideally, work out of the box.

Aims to support [#48454](https://github.com/tensorflow/tensorflow/issues/48454), [#25218](https://github.com/tensorflow/tensorflow/issues/25218),  in the future
Fixes [#17097](https://github.com/tensorflow/tensorflow/issues/17097)

### CODE REVIEW

**Comment by DarshanDeshpande** [comment]
@gbaned Any updates on this?

**Comment by bhack** [comment]
@DarshanDeshpande What is missing to full fix https://github.com/tensorflow/tensorflow/issues/17097 ?

**Comment by DarshanDeshpande** [comment]
@bhack It's fully fixed now. I forgot to update the ones initializer in my first commit so I edited the main message to partially fixes but forgot to edit it again when I merged the second commit. I have changed it to fixes. Thanks :)

**Comment by bhack** [comment]
Thanks for your first contribution.
We will assign that ticket to you and we will move it `in progress`.  
In the meantime If you are interested in other contributions please check also the TODO column in https://github.com/orgs/tensorflow/projects/14. 
Comment on the ticket that you like to work on and we will assign it to you before the PR so that we will not have potentially duplicated coding activities.

**Comment by DarshanDeshpande** [comment]
Sorry I closed this PR accidentally. @deeb02 @bhack @gbaned Could you help me with the failed tests, please? I have fixed the NoneType error that popped up in the previous test but I am not sure how to fix the rest since tf.keras isn't fully compatible with complex numbers at the moment and keras.backend.variable requires the weights to be of floating data type (which is causing some tests to fail)

**Comment by DarshanDeshpande** [comment]
@bhack @gbaned @deeb02 I have fixed the failing tests. Please check and let me know if any other changes are to be made.

**Review by bhack — APPROVED** [review]
Just to run tests

**Comment by DarshanDeshpande** [comment]
@gbaned @deeb02 Any update yet?

**Comment by DarshanDeshpande** [comment]
Thanks for the review @mattdangerw! The initial idea for this pull request was to help expand complex number support for tf.keras. 

If you take a look at this [colab ](https://colab.research.google.com/drive/1zWAvg3KbLUflAvKPPsWaR4dJvqugYL5-?usp=sharing) notebook, you can see that if the `Input` layer is of type tf.complex64/128 then the `Dense` layer weights following it are automatically cast to float (depending on backend.floatxx). If you try to specify a custom initializer, it throws an error saying complex dtype is not supported. A simple dense layer can be implemented which works with complex numbers since tf.matmul deals with complex numbers just fine (see example in the notebook). Initially, when I created this PR, I thought that these errors were because of a lack of support for weight initializers but then it looks like `tensorflow/core/ops` don't support complex numbers for most layers including Dense (again, tf.matmul op supports complex values) and Convolution (see notebook for a possible implementation example). I am working on creating a couple more PRs to add support for these ops as we speak but till then the addition of proper support for all complex weight initializers to test those new changes would make it a bit easier to work with and will be useful once the layers are made compatible. Keras already has optimizer support for complex gradients and so this PR will act as a small step towards achieving complete complex variable support, which is the bigger picture you spoke about :)

**Comment by DarshanDeshpande** [comment]
@mattdangerw If there are any thoughts or suggestions regarding the same or if there are any updates related to this PR then please let me know!

**Comment by mattdangerw** [comment]
Thanks for the response!

I guess maybe a better specific question is what do the changes to the ones, zeros and constant intializers aim to accomplish? You can call these today with complex dtypes and they work. You can use these in a Dense layer and they seem to work:
`tf.keras.layers.Dense(40, kernel_initializer='ones', dtype=tf.complex64)`

I can see the point about a default initializer that is inferred for in `add_weight` not handling complex dtypes (the note you linked), but it looks like really the way this is solving that issue is by supporting complex number with glorot_uniform.

I am also a little confused at why these changes seem to initialize everything with an equivalent real and imaginary value. That does not seem to be correct--I would expect ones to return `1 + 0i` as it does today. I would also expect random initializers to not constraint the real and imaginary part to be equal, but that could definitely be my own ignorance on the subject matter. Do we have references here?

Finally, there is a layering issue. We are working to remove all private usage of lower level tf APIs from keras.  `gen_math_ops._complex` would qualify, we should avoid that usage.

**Comment by DarshanDeshpande** [comment]
Thanks for the clarification @mattdangerw!

1. I haven't touched the zeros and constant initializers since they already support complex numbers. Regarding the ones initializer, the distribution of ones should logically be in both real and imaginary (or amplitude and phase components of the weights) instead of just the real domain. That is the reason why the ones initializer was updated in later commits.

2. The note that I referred to was just to show that currently, by default, the code throws an error if the initializer is purposely set to None. This PR will add support for Glorot, He and Lecun initializers and so you need not worry about it as it will be handled on its own.

3. You brought up a very important question that I was meaning to ask. I had even mailed about 10-15 members from the Keras and Tensorflow team but got no response. The weight initialization that is implemented follows the intuition that both the real and imaginary components should have same distributions since both are in separate vector spaces. I had done my research on this and I could only find [this](https://arxiv.org/abs/1705.09792v1) paper whose initializer code is present [here](https://github.com/ChihebTrabelsi/deep_complex_networks/blob/f7065cea654c8b2a049c40fcc40fd76817c15096/complexnn/init.py#L202). This implementation uses Rayleigh distribution but I couldn't implement it because the only rayleigh functionality I could find for tensorflow was present in TFP ([here](https://www.tensorflow.org/probability/api_docs/python/tfp/random/rayleigh)) which is a completely separate branch of Tensorflow. Most other papers tend to use zeros or ones or something similar, so I followed the next best implementation according to me which was having the same distributions (uniform/normal) in both domains. About the usage of same seed values for both domains, this can be changed if you want but I'm not sure if it will have any significant improvement in terms of providing better loss convergence.

4.  I didn't know about the layering issue. Thanks for letting me know! I will change it to `math_ops.complex` instead

**Comment by mattdangerw** [comment]
I do not see a good reason why the ones initializer should have different behavior than `tf.ones` with a complex dtype. If you want `1 + i` you could use a constant initializer right? `1 + i` is not even norm one. I think this will surprise people.

Re point 3) having the same distribution does mean having equal value I think. We should be careful to make sure we are doing the right thing here, but sadly I'm probably not the best person to ask. There is a project out there trying to do this  https://github.com/NEGU93/cvnn/blob/master/cvnn/initializers.py. It looks like some care was taken to handle the limits of the distribution differently, and it looks like will only return a non-zero real component. I will try to ask around to find if someone has done more thinking on this.

**Comment by DarshanDeshpande** [comment]
@mattdangerw I have updated the `Ones` initializer and changed to `math_ops.complex` instead of `gen_math_ops` as per your advice. 

The repository that you linked to has a similar weight initialization implementation, but with different seeds for real and imaginary kernels. Rather than initializing complex-valued weights, the author initializes two real-valued kernels, one for each real and imaginary domains and handles their operations separately. So every time `add_weight` is called, it results in two different-valued real and imaginary domain arrays. This is the current best way of handling complex numbers in tensorflow without changing the C++ ops but is inefficient since it doubles the cost of operations like convolution (which as I had mentioned in my notebook can be done much more efficiently by using complex-valued weights). 

Let me know if you think we should use different seeds for initialization

**Comment by deeb02** [comment]
Hey @DarshanDeshpande, I wanted to take a step back and ask about the motivation. Is it neural networks use case or something else? Is it to be able to handle special types of inputs for some problem domains? I took a look at #48454, #25218 but still couldn't figure out the main use case.

**Comment by DarshanDeshpande** [comment]
@deeb02 Okay so if you're familiar with the signal and audio processing domains, you probably know about the widespread use of complex numbers in the field. I will provide a very basic example to explain the need for this PR:
If anyone wants to train a ML model for an audio task, whether it be classification or generation based, the usual method is to use a FFT, STFT or wavelet based transformation to convert the data into a time-frequency representation. These methods project the timeseries data into complex domains and this data is usually converted to real spectrograms which causes loss of individuality of phase and magnitude that is irrecoverable. To avoid this, we usually split the STFT output as two real tensors and stack them up, thereby changing their shape to (2, frames, freq bins) instead of the original (frames, freq bins). This is extremely wasteful and feeds the dimensionality curse. Also, as I had mentioned in my conversation with @mattdangerw, the current best way to process this is to use two kernels, one for each component which is computationally expensive. The best and the most efficient way is to use complex numbers directly while training since complex gradients are already supported. But to be able to do this, we will need to expand tensorflow's support for complex numbers and this PR will act as a starting step for that by adding complete weight initializer support, upon which the support for individual layers can be added.

**Comment by fchollet** [comment]
Thank you for the PR!

In general, we want to make sure that use cases that involve complex numbers are *possible* in Keras, but we have intention to support them *out of the box* given that this is a very niche space.

Basically, it should be possible for you to easily roll out your own complex-valued initializers and complex-valued layers, and have them work. However, we don't intend to add complex support to existing real-valued initializers and layers.

In the case of this PR, I recommend that you create your own `Initializer` subclasses that return complex values. If there's a significant amount of interest in such initializers, then we can consider making them available in the TF Addons package.

**Comment by bhack** [comment]
/cc @nikitamaia we have still the other two mentioned issues in https://github.com/tensorflow/tensorflow/pull/48724#issue-622246567 as `contribution welcome`.
Probably it Is better to evaluate/close these before another PR will land in the repo.
